feat: Add workspace agent for SSH

kylecarbs opened the pull request:
This adds the initial agent that supports TTY
and execution over SSH. It functions across MacOS,
Windows, and Linux.

This does not handle the coderd interaction yet,
but does setup a simple path forward.

kylecarbs committed:
feat: Add workspace agent for SSH

This adds the initial agent that supports TTY
and execution over SSH. It functions across MacOS,
Windows, and Linux.

This does not handle the coderd interaction yet,
but does setup a simple path forward.

bryphe-coder reviewed (approved):
Just some questions inline - but overall looks great. Exciting to have the `agent` piece come together. Hopefully that means we can `ssh` into workspaces soon 😄

kylecarbs committed:
Merge branch 'main' into workspaceagent

kylecarbs committed:
Fix pty tests on Windows

kylecarbs committed:
Fix log race

kylecarbs committed:
Lock around dial error to fix log output

kylecarbs committed:
Fix context return early

kylecarbs committed:
fix: Leaking yamux session after HTTP handler is closed

Closes #317. We depended on the context canceling the yamux connection,
but this isn't a sync operation. Explicitly calling close ensures the
handler waits for yamux to complete before exit.

kylecarbs committed:
Merge branch 'main' into closeconn

kylecarbs committed:
Lock around close return

kylecarbs committed:
Merge branch 'main' into closeconn

kylecarbs committed:
Force failure with log

kylecarbs committed:
Merge branch 'main' into closeconn

kylecarbs committed:
Merge branch 'closeconn' into workspaceagent

kylecarbs committed:
Fix failed handler

kylecarbs committed:
Upgrade dep

kylecarbs committed:
Fix defer inside loops

kylecarbs committed:
Fix context cancel for HTTP requests

kylecarbs committed:
Merge branch 'main' into workspaceagent

kylecarbs committed:
Fix resize
